### DISCUSSION TITLE
PDF to image with optional content

### DISCUSSION THREAD

**Discussion by cfcurtis** [pymupdf/PyMuPDF#2136] [Looking for help] [answered]
Hello! I'm looking for a way of converting a PDF to either PNG or SVG one layer at a time. I know that OCGs are tricky business in PDFs, but I was hoping that PyMuPDF might have this functionality already.

I have tried the following, for a document `doc` that has only 2 pages but ~20 OCGs:

```python
for xref, ocg_dict in doc.get_ocgs().items():
    doc.set_layer(-1, basestate="OFF", on=[xref])
    for page in doc.pages():
        dl = page.get_displaylist()
        pix = dl.get_pixmap()
        pix.save(f"page{page.number}-{ocg_dict['name']}.png")
```
This gave me 20 identical images per page, with each showing the content of all layers.

I also tried modifying the visibility with `set_layer_ui_config` and had the same outcome.

I'm guessing layer visibility isn't taken into account with the underlying `get_pixmap` implementation, but I thought I'd ask before I start parsing the content stream myself. Thanks!

**Comment by JorjMcKie** [comment, marked as answer]
Can you let me have that PDF?
You sure that the OCGs really imply different views?

**Reply by cfcurtis** [reply]
Thank you for the quick reply! I am working with sewing patterns, which typically have each size drawn in a different OCG. These are proprietary documents, but here's a single page that I have obfuscated to remove identifying information.
[b548aca01e1be9cba9bb2c1bad29ca8b.pdf](https://github.com/pymupdf/PyMuPDF/files/10247425/b548aca01e1be9cba9bb2c1bad29ca8b.pdf)

Edit: I'm not sure what you mean by different views, but enabling/disabling layer visibility in a PDF viewer does show/hide different elements.

**Reply by JorjMcKie** [reply]
The following works:
```python
doc=fitz.open(doc.name)
ui_configs=doc.layer_ui_configs()  # all the different ui layers
page=doc[0]
for i in range(len(ui_configs)):  # walk through all layers and toggle visibility of each
    pix=page.get_pixmap()
    pix.save(f"{i}.png")
    doc.set_layer_ui_config(i, action=1) # toggle on <-> off for next pixmap

pix = page.get_pixmap()  # image of final layer setting
pix.save(f"{i+1}.png")
```
So i=0 images the state as stored in document, subsequent iterations toggle visibility of layers one by one.

**Reply by cfcurtis** [reply]
Thank you! This is working, I was giving set_layer_ui_config the xref instead of the layer index.

My final version to save each layer independently is a bit ugly, but it works:

```python
ui_configs = doc.layer_ui_configs()
[doc.set_layer_ui_config(i, action=2) for i in range(len(ui_configs))] # turn off all layers
for layer in ui_configs:
    doc.set_layer_ui_config(layer["number"], action=0) # turn on the current layer
    
    for page in doc:
        pix = page.get_pixmap()
        pix.save(f"page{page.number:02d}-layer{layer['text']}.png")
    
    doc.set_layer_ui_config(layer["number"], action=2) # turn off the current layer
```

**Comment by JorjMcKie** [comment]
@cfcurtis - just wondered why you call your solution "ugly". Maybe you wanted PDF results that didn't have images as pages? But instead pages in original PDF format?
If yes, then let me confirm that this also is possible:
You only need to deactivate unwanted OCG(s) via `Document.set_layer()`, then save the PDF, then close-open and repeat with the next OCGs, ...

**Reply by cfcurtis** [reply]
I just meant that the code is a bit ugly as I'm looping through all the layers twice, the behaviour is exactly what I wanted it to do :).

I actually did some other work with PDFs where I wanted to fully remove optional content (not just deactivate), and I ended up having to parse the content stream and strip out drawing operations in the unwanted OCGs.

**Comment by NurielWainstein** [comment]
Hi, how are you?
is it possible to export the files in pdf format instead of image?
so each pdf will contain only one layer.

**Reply by cfcurtis** [reply]
When you say "contain only one layer", do you mean only have one layer visible, or actually remove the layer content entirely? Visibility is easy - as with @JorjMcKie's comment, you can simply call [Document.set_layer()](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_layer) and then save the document.

To actually *remove* content entirely, that's a whole different story. I do it [here](https://github.com/cfcurtis/pdfstitcher/blob/main/pdfstitcher/processing/layerfilter.py) (using PikePDF, but PyMuPDF could do it too) but it involves parsing content streams (often recursively) and keeping track of the graphics state - you can't simply delete content in a BMC/EMC block and expect to have the same visuals in the stuff that's left.

**Reply by NurielWainstein** [reply]
about the first - I tried saving after leaving only one layer with active visibility wiht doc.save("o.pdf") but I'm still seeing all the layers...

about the second(which is the preferred way), do you have a code example that given "path/input.pdf" I'll get the required behavior? 

anyway, thank you very much, didn't expect to actually get a response haha...

**Reply by cfcurtis** [reply]
@NurielWainstein I'm just surprised this is almost a 2 year old discussion!

If you show your code then maybe we can figure out why the visibility saving isn't working.

If you want to try PDFStitcher, it's not very well-documented at the moment but it'd be something like:

```bash
pip3 install pdfstitcher
pdfstitcher path/input.pdf
```

This will list the layers present in the document, along with some other info. Then, you can select which layer(s) to keep with the following command:

```bash
pdfstitcher -k "Layer name" path/input.pdf path/output.pdf
```

For multiple layers it's looking for a semicolon separated list (comma seemed too likely to be in an actual layer name), e.g. `"Layer 1;Layer 2;Layer 3"`

@JorjMcKie I apologize for taking this discussion away from PyMuPDF!

**Reply by NurielWainstein** [reply]
your code works perfectly! so this pretty much solved my problem.

about the first option, Im just running doc.save("out.pdf") inside the loop...
```
import fitz
input = "pdfs/test1.pdf"

doc = fitz.open(input)

ui_configs = doc.layer_ui_configs()
[doc.set_layer_ui_config(i, action=2) for i in range(len(ui_configs))]
for layer in ui_configs:
    if "EX-WALL" in layer['text']:
        doc.set_layer_ui_config(layer["number"], action=0)
        doc.save("out.pdf")
        for page in doc:
            pix = page.get_pixmap()
            pix.save(f"page{page.number:02d}-layer{layer['text']}.png")

        doc.set_layer_ui_config(layer["number"], action=2)

```
input file:
[test1.pdf](https://github.com/user-attachments/files/15909254/test1.pdf)
